File: 3105. Longest Strictly Increasing or Strictly Decreasing Subarray/javascript/main.js

// Brute force
var convertToDirection = (num) => {
  return num > 0 ? 1 : num < 0 ? -1 : num;
};
/**
 * @param {number[]} nums
 * @return {number}
 */

var longestMonotonicSubarray = function (nums) {
  if (nums.length <= 1) {
    return nums.length;
  }

  let previousDirection = convertToDirection(nums[1] - nums[0]);
  let result = previousDirection !== 0 ? 2 : 1;
  let tempSum = previousDirection !== 0 ? 2 : 1;

  for (let index = 2; index < nums.length; index++) {
    let currentDirection = convertToDirection(nums[index] - nums[index - 1]);
    if (currentDirection === 0 || currentDirection !== previousDirection) {
      result = Math.max(tempSum, result);
      tempSum = currentDirection !== 0 ? 2 : 1;
    } else {
      tempSum += 1;
    }
    previousDirection = currentDirection;
  }

  return Math.max(tempSum, result);
};

// Two counters (more straight-forward)

var longestMonotonicSubarray2 = function(nums) {
    if (nums.length <= 1) return nums.length;
    
    let inc = 1, dec = 1, maxLen = 1;
    
    for (let i = 1; i < nums.length; i++) {
        if (nums[i] > nums[i-1]) {
            inc++;
            dec = 1;
        } else if (nums[i] < nums[i-1]) {
            dec++;
            inc = 1;
        } else {
            inc = 1;
            dec = 1;
        }
        maxLen = Math.max(maxLen, inc, dec);
    }
    
    return maxLen;
};

console.log(longestMonotonicSubarray([3, 4, 7, 7]));

File: 3105. Longest Strictly Increasing or Strictly Decreasing Subarray/python/main.py
"""Module to find the longest strictly increasing or strictly decreasing subarray."""

from typing import List

class Solution:
    """Class to find the longest monotonic subarray."""  
    def longest_monotonic_subarray(self, nums: List[int]) -> int:
        """Return the length of the longest strictly increasing or strictly decreasing subarray."""
        if len(nums) <= 1:
            return len(nums)
        inc, dec, max_len = 1, 1, 1
        for i in range(1, len(nums)):
            if nums[i] > nums[i-1]:
                inc = inc + 1
                dec = 1
            elif nums[i] < nums[i-1]:
                dec = dec + 1
                inc = 1
            else:
                inc = 1
                dec = 1
        max_len = max(max_len, inc, dec)
        return max_len

if __name__ == "__main__":
    s = Solution()
    print(s.longest_monotonic_subarray([1,4,3,3,2]))
    